### PR TITLE
[JsonSerializer] Add support for out-of-order metadata reads.

### LINKED ISSUE #72604: JsonSerializer: Allow out-of-order reading of metadata properties.
**EDIT** See https://github.com/dotnet/runtime/issues/72604#issuecomment-1904569859 for an API proposal.

### Description

Attempting to deserialize a polymorphic structure with System.Text.Json that doesn't feature `$type` at start of the object results in an exception.

For my scenario, this currently prevents me from using the built-in polymorphism support with `jsonb` columns in Postgres, as object properties have no guaranteed order.

Developer comment on this issue: https://github.com/dotnet/runtime/issues/63747#issuecomment-1158624112

### Reproduction Steps

```csharp
var databaseState = @"{
  ""BaseDictionary"": {
    ""6ed6e524-2ca4-4fea-8e21-7245ccb61863"": {
      ""Id"": ""6ed6e524-2ca4-4fea-8e21-7245ccb61863"",
      ""Name"": ""Something"",
      ""$type"": ""Derived"",
      ""OtherGuid"": ""d471c77d-5412-4e7a-a98d-8304e87792ed""
    }
  }
}";

JsonSerializer.Deserialize<WrapperType>(databaseState);

public record WrapperType(Dictionary<Guid, WrapperType.Base> BaseDictionary)
{
    [JsonDerivedType(typeof(Derived), nameof(Derived))]
    [JsonDerivedType(typeof(AlsoDerived), nameof(AlsoDerived))]
    public abstract record Base(Guid Id, string Name);
    public record Derived(Guid Id, string Name, Guid OtherGuid): Base(Id, Name);
    public record AlsoDerived(Guid Id, string Name) : Base(Id, Name);
}
```

### Expected behavior

Deserialization should work.

### Actual behavior

```
System.NotSupportedException: 'Deserialization of types without a parameterless constructor, a singular parameterized constructor, or a parameterized constructor annotated with 'JsonConstructorAttribute' is not supported. 
```

### Regression?

Limitation of initial implementation

### Known Workarounds

I currently use [PolyJson](https://github.com/rmja/PolyJson) as an alternative, as the implementation reads ahead to find the discriminator.

### Configuration

Impacts any version of STJ 7.0

### Other information

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Makes the following changes:

1. Adds support for out-of-order metadata reads in JsonSerializer.
2. Ensures that read JSON property names are always unescaped.

Fix #72604.

### CODE REVIEW

**Comment by eiriktsarpalis** [comment]
New API pending approval, please do not merge yet.

File: src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs:1517]: Updates the test to account for property unescaping now taking place.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:284]: Why are these ifdefs needed? Inlining doesn't happen in debug, anyway.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:284]: It doesn't? I've never been able to hit breakpoints in these methods.

File: src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs:46]: > which would (correct me if I'm wrong) normally be an error

It might not be, for example it's possible to specify a type discriminator for a derived type but not the base type so its presence on the wire can be entirely optional.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs:46]: That is to say, we might decide in API review that we want this to be the default behavior (at the risk of overbuffering or even OOM'ing for scenaria that used to work in the past)

Comment (stephentoub) [reply src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs:46]: > > which would (correct me if I'm wrong) normally be an error
> 
> It might not be, for example it's possible to specify a type discriminator for a derived type but not the base type so its presence on the wire can be entirely optional.

I still don't see how that affects the answer. If the necessary type discriminator is first, today it's valid, there's no buffering, and tomorrow there wouldn't be buffering either. If the necessary type discrimator is not first, today that's an error, and tomorrow there would be buffering but it would only be in cases where it was erroneous already; it wouldn't be making something that's currently fast instead be slow.

Where am I getting that wrong?

File: src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs
Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs:1517]: What's the motivation for now unescaping the metadata properties? IIRC, we choose not to, in order to use escaping as a workaround for users that wanted to use ReferenceHandler.Preserve with the reserved properties.

See https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md#json-objects-if-not-enumerable-class--struct--dictionary---on-deserialize-and-serialize.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/PolymorphicTests.CustomTypeHierarchies.cs
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/PolymorphicTests.CustomTypeHierarchies.cs:2334]: I was wondering if this was a valid scenario, cool 😄

Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/PolymorphicTests.CustomTypeHierarchies.cs:2334]: Does this combination work with a custom type discriminator?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs
Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:284]: @EgorBo, am I wrong?

Comment (EgorBo) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:284]: Yes, it shouldn't happen in Debug

File: src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs:46]: > If the necessary type discriminator is first, today it's valid, there's no buffering, and tomorrow there wouldn't be buffering either. 

It would buffer even in that scenario since it would want to look ahead for other metadata properties such as `$id` or `$values`.

> Where am I getting that wrong?

Consider the hierarchy:

```C#
[JsonDerivedType(typeof(Derived), "derived"))]
public record Base(IEnumerable<int> values);
public record Derived(int x) : Base([x]);
```

This makes type discriminators legal because you have Derived but also optional since `Base` doesn't specify one. In that case you could today write something like
```C#
var value = new Base(Enumerable.Repeat(1, int.MaxValue));
using fs = File.OpenWrite("/some/path");
JsonSerializer.Serialize(fs, value); // { "values" : [1, 1, ... ] }
```
Currently the above can serialize and deserialize without any overbuffering whatsoever, however assuming we turned on reading ahead by default then it would try to buffer the entire payload so that it can look for the optional `$type` declaration.

File: src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs:1517]: Thanks, I had assumed this was intentional but I hadn't seen that doc.

Couple of reasons, since polymorphism metadata property names can be user-defined (and therefore assuming the property name is escaped it would be impossible to round-trip). Secondly, it's inconsistent with how property names are handled elsewhere so that workaround happens to work for reference preservation but it doesn't for regular property binding.

I'd be curious to hear your thoughts on the matter, but I'd be inclined to remove this behavior (and perhaps file a breaking change issue).

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/PolymorphicTests.CustomTypeHierarchies.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/PolymorphicTests.CustomTypeHierarchies.cs:2334]: It should do, I'll see if I can add a similar test for a type like that.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:96]: The old message said this:

> The metadata property '$id' must be the first reference preservation property in the JSON object.

Technically this hasn't been accurate since polymorphism got introduced, since following `$id` after `$type` is supported. This narrows down the message to the actual failure mode it is still being used for.

File: src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs
Comment (stephentoub) [reply src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs:46]: And that must be the behavior, that the last rather than first discriminator is used? If they're "allowed out of order", isn't that don't order doesn't matter and why can pick whichever?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs:46]: "out of order" in this context means making it possible to intersperse metadata properties among data. It doesn't have any impact on how duplicate properties are handled, in both modes an exception will be thrown eventually.

Per the example above, the problem comes from having to buffer when reading a valid encoding that doesn't have any metadata in its payload.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs:46]: It's by design. The presence of a type discriminator is optional (`JsonDerivedType` attribute annotations don't require it). This is to cater to polymorphic scenaria where emitting discriminator metadata is undesirable (and round tripping is not a concern). We could have required that annotations be consistent for a given hierarchy, but ultimately this can only be verified at run time and is less flexible (e.g. it would require all types in the open hierarchy to be annotated, including the base type itself). 

Ultimately though, there are more metadata properties than just `$type`. We'd still want to look ahead for potential `$values` for cases where some of the types in the hierarchy are collections, or `$id` for cases where we also have reference preservation enabled.

File: src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs
Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs:1517]: > it's inconsistent with how property names are handled elsewhere

Yes, it was inconsistent on purpose because metadata properties are not meant to be treated as regular properties by the serializer, the Utf8JsonReader does treat them normally though.

It's probably OK to change but my current understanding is that no user is complaining about this, is that correct?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs:1517]: No user-reported issues just yet, but it is for example failing to deserialize the following example correctly:

```C#
Base result = JsonSerializer.Deserialize<Base>("""{"naïve" : "derived" }""");
Console.WriteLine(result is Derived); // False

[JsonPolymorphic(TypeDiscriminatorPropertyName = "naïve")]
[JsonDerivedType(typeof(Derived), "derived")]
public record Base;
public record Derived : Base;
```

I can revert this if you feel that it's out of scope, but this seems to be a bug in my view.

Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs:1517]: No need to revert, it looks like a change for the better, assuming a breaking change is filed. We also have plenty of time to expect customer reports in case someone is affected. Thanks for explaining and sharing that scenario.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.Json/src/Resources/Strings.resx
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/Resources/Strings.resx:389]:
```suggestion
    <value>The JSON payload for polymorphic interface or abstract type '{0}' must specify a type discriminator.</value>
```

**Pull request merged**
